llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,447 @@
+//===- SYCLBIN.cpp - SYCLBIN binary format support --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Object/SYCLBIN.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Object/OffloadBinary.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace llvm::object;
+
+namespace {
+
+class SYCLBINBlockReader {
```

Comment:
nit.
These 3 classes could be replaced with simple functions making the maintenance easier for developers.

Reply:
I think the benefit of the class is that we can implicitly track how much memory is left. This lets us avoid reading outside the buffer, which is good for security. We could of course track that through functions as well, but it would require some additional variables we pass around. Personally I think this is cleaner, but if you think having free-functions for it is better, I won't die on this hill. 😉